File: crates/dewasm-backend-java/units/rt/tail_call.java
// The trampoline a tail-calling function's entry runs.
// A tail call parks its target and arguments on the instance and returns; nothing is allocated per hop, which is what a chain of any length pays otherwise.
// A tail entry is bound to the instance that built it and reads *that* instance's parked slots, which is why `table/tail_slot` hands one back only to its owner.
interface TailBody {
	Object run();
}

File: crates/dewasm-backend-ruby/units/rt/tail_call.rb
# A pending tail call: a tail-calling function's body returns one of these instead of calling, and its entry wrapper unwraps them in a loop, so a chain of any length runs in constant stack space.
TailCall = Struct.new(:target, :args)

def trampoline(r)
  r = r.target.call(*r.args) while r.is_a?(TailCall)
  r
end
